File: js/test_money.js
const assert = require("assert");
const Money = require("./money");
const Portfolio = require("./portfolio");
const Bank = require("./bank");

class MoneyTest {
  setUp() {
    this.bank = new Bank();
    this.bank.addExchangeRate("EUR", "USD", 1.1);
    this.bank.addExchangeRate("USD", "KRW", 1100);
  }

  testMultiplication() {
    let fiveDollars = new Money(5, "USD");
    let tenDollars = fiveDollars.times(2);
    let expectedUsd = new Money(10, "USD");

    assert.deepStrictEqual(tenDollars, expectedUsd);
  }

  testDivision() {
    let originalMoney = new Money(4002, "KRW");
    let actualMoneyAfterDivision = originalMoney.divide(4);
    let expectedMoneyAfterDivision = new Money(1000.5, "KRW");

    assert.deepStrictEqual(
      actualMoneyAfterDivision,
      expectedMoneyAfterDivision
    );
  }

  testAddition() {
    let fiveDollars = new Money(5, "USD");
    let tenDollars = new Money(10, "USD");
    let fifteenDollars = new Money(15, "USD");
    let portfolio = new Portfolio();
    portfolio.add(fiveDollars, tenDollars);
    assert.deepStrictEqual(
      portfolio.evaluate(this.bank, "USD"),
      fifteenDollars
    );
  }

  testAdditionOfDollarsAndEuros() {
    let fiveDollars = new Money(5, "USD");
    let tenEuros = new Money(10, "EUR");
    let portfolio = new Portfolio();
    portfolio.add(fiveDollars, tenEuros);
    let expectedValue = new Money(16, "USD");
    assert.deepStrictEqual(portfolio.evaluate(this.bank, "USD"), expectedValue);
  }

  testAdditionOfDollarsAndWons() {
    let oneDollar = new Money(1, "USD");
    let elevenHundredWon = new Money(1100, "KRW");
    let portfolio = new Portfolio();
    portfolio.add(oneDollar, elevenHundredWon);
    let expectedValue = new Money(2200, "KRW");
    assert.deepStrictEqual(portfolio.evaluate(this.bank, "KRW"), expectedValue);
  }

  testAdditionWithMultipleMissingExchangeRates() {
    let oneDollar = new Money(1, "USD");
    let oneEuro = new Money(1, "EUR");
    let oneWon = new Money(1, "KRW");
    let portfolio = new Portfolio();
    portfolio.add(oneDollar, oneEuro, oneWon);
    let expectedError = new Error(
      "Missing exchange rate(s):[USD->Kalganid,EUR->Kalganid,KRW->Kalganid]"
    );
    assert.throws(
      () => portfolio.evaluate(this.bank, "Kalganid"),
      expectedError
    );
  }

  testConversionWithDifferentRatesBetweenTwoCurrencies() {
    let bank = new Bank();
    bank.addExchangeRate("EUR", "USD", 1.2);
    let tenEuros = new Money(10, "EUR");
    assert.deepStrictEqual(bank.convert(tenEuros, "USD"), new Money(12, "USD"));

    this.bank.addExchangeRate("EUR", "USD", 1.3);
    assert.deepStrictEqual(
      this.bank.convert(tenEuros, "USD"),
      new Money(13, "USD")
    );
  }

  testConversionWithMissingExchangeRate() {
    let bank = new Bank();
    let tenEuros = new Money(10, "EUR");
    let expectedError = new Error("EUR->Kalganid");
    assert.throws(function () {
      bank.convert(tenEuros, "Kalganid");
    }, expectedError);
  }

  getAllTestMethods() {
    let moneyPrototype = MoneyTest.prototype;
    let allProps = Object.getOwnPropertyNames(moneyPrototype);
    let testMethods = allProps.filter((p) => {
      return typeof moneyPrototype[p] === "function" && p.startsWith("test");
    });
    return testMethods;
  }

  runAllTests() {
    let testMethods = this.randomizeTestOrder(this.getAllTestMethods());
    testMethods.forEach((m) => {
      console.log("Running: %s()", m);
      let method = Reflect.get(this, m);
      try {
        this.setUp();
        Reflect.apply(method, this, []);
        console.log("Pass");
      } catch (e) {
        if (e instanceof assert.AssertionError) {
          console.log(e);
        } else {
          throw e;
        }
      }
    });
  }

  randomizeTestOrder(testMethods) {
    for (let i = testMethods.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [testMethods[i], testMethods[j]] = [testMethods[j], testMethods[i]];
    }
    return testMethods;
  }
}

new MoneyTest().runAllTests();

File: py/test_money.py
import unittest
from money import Money
from portfolio import Portfolio


class TestMoney(unittest.TestCase):
  def testMultiplication(self):
    fiver = Money(5, "USD")
    tenner = fiver.times(2)
    expected_usd = Money(10, "USD")
    self.assertEqual(expected_usd, tenner)

  def testMultiplicationInEuros(self):
    ten_euros = Money(10, "EUR")
    twenty_euros = ten_euros.times(2)
    expected_eur = Money(20, "EUR")
    self.assertEqual(expected_eur, twenty_euros)

  def testDivision(self):
    original_money = Money(4002, "KRW")
    actual_money_after_division = original_money.divide(4)
    expected_money_after_division = Money(1000.5, "KRW")
    self.assertEqual(expected_money_after_division,
                      actual_money_after_division)

  def testAddition(self):
    five_dollars = Money(5, "USD")
    ten_dollars = Money(10, "USD")
    fifteen_dollars = Money(15, "USD")
    portfolio = Portfolio()
    portfolio.add(five_dollars, ten_dollars)
    self.assertEqual(fifteen_dollars, portfolio.evaluate("USD"))

  def testAdditionOfDollarsAndEuros(self):
    five_dollars = Money(5, "USD")
    ten_euros = Money(10, "EUR")
    portfolio = Portfolio()
    portfolio.add(five_dollars, ten_euros)
    expected_value = Money(16, "USD")
    actual_value = portfolio.evaluate("USD")
    self.assertEqual(expected_value, actual_value, "%s != %s" % (expected_value, actual_value))

  def testAdditionOfDollarsAndWons(self):
    one_dollar = Money(1, "USD")
    eleven_hundred_won = Money(1100, "KRW")
    portfolio = Portfolio()
    portfolio.add(one_dollar, eleven_hundred_won)
    expected_value = Money(2200, "KRW")
    actual_value = portfolio.evaluate("KRW")
    self.assertEqual(expected_value, actual_value, "%s != %s" % (expected_value, actual_value))

  def testAdditionWithMultipleMissingExchangeRates(self):
    one_dollar = Money(1, "USD")
    one_euro = Money(1, "EUR")
    one_won = Money(1, "KRW")
    portfolio = Portfolio()
    portfolio.add(one_dollar, one_euro, one_won)
    with self.assertRaisesRegex(
      Exception,
      "Missing exchange rate\(s\):\[USD\->Kalganid,EUR->Kalganid,KRW->Kalganid]",
    ):
      portfolio.evaluate("Kalganid")

if __name__ == '__main__':
    unittest.main()